Makefile colorizing improvement: colorize at the beginning of line, handling the '\' symbol and other.

fadeevab opened the pull request:
1). Colorize when the expression at the beginning of line.
2). Handle '\\' at the end of prerequisites (the next line should be colorized as a continuation of previous).
3). Colorize (origin|flavor).
4). Fix colorizing of built-in which stands after a common variable.

![makefile_before](https://cloud.githubusercontent.com/assets/5967447/26456340/7136beea-4175-11e7-985b-d3d6d7af3273.png)
![makefile_after](https://cloud.githubusercontent.com/assets/5967447/26456339/71335d90-4175-11e7-8d47-4c677847743f.png)

**Note:** tested only at JSON version in VS Code. Current patch is ported from JSON to XML and is not verified. I appreciate anybody could help to test this patch. Thank you!

fadeevab commented:
Could somebody review? Thank you!

fadeevab commented:
@sorbits, @alkemist, @bradchoate, @infininight, As I see your names in TextMate project, could anybody pay attention to this pull request? Thank you!

infininight commented:
@fadeevab I'll be the one reviewing the pull request. This may a little while however as Makefiles to no have any language spec that I've ever found. So deducing what is allowed and where is a bit hit or miss. There are also a few other issues I am already aware of so I will likely do a larger grammar review as I review the PR.

fadeevab commented:
@infininight , It could be helpful that there are some kind of syntax blocks in https://www.gnu.org/software/make/manual/make.html, e.g. for variable:

```
immediate = deferred
immediate ?= deferred
immediate := immediate
immediate ::= immediate
immediate += deferred or immediate
immediate != immediate
```
It's spread through the document though.

fadeevab commented:
Could just make simple test to check makefile is colorizing with this patch and, possibly, approve? My point: I assume you have no enough time to make detailed review, while the current patch doesn't make colorizing worse, and even after commit more cases handling is needed, and I will push more PR. So need to move forward. Thank you for your attention.

fadeevab commented:
@infininight, ...just reminding about review. Thanks

fadeevab commented:
@sorbits, @alkemist, @bradchoate, @infininight, Hi, there! Initially I prepared the pull request into VS Code repository, but VS Code team refuses my patch until TextMate approves current pull request. Reason: to not create individual fork of colorizing, and this is a good point. Could you, please, review and approve this patch? Thank you!

infininight commented:
Can you break this into separate discrete commits? Want to separate the changes into their purpose. A couple aren't making sense.

fadeevab commented:
Ok, I can.

fadeevab commented:
@infininight, Could you, please, suggest me how to test this XML in Ubuntu? Thank​ you!

infininight commented:
The only thing I am aware of is the Lightshow tester from Github:

https://github-lightshow.herokuapp.com

This is limited though as it only shows a subset of scopes and there is no way to see them directly.

If it's an issue I can piece them out myself but will need to ask some questions likely to do so.

fadeevab commented:
@infininight, Thanks, I'll check it.

fadeevab commented:
@infininight , Finally I found the time to check https://github-lightshow.herokuapp.com, and unfortunately, it doesn't even react for my custom requests, it just stuck at the GET request.
Shortly, a huge probability, that I can't check and divide the patch, and thus this patch is stuck :(

fadeevab commented:
Checked with their tests [here](https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_url=https%3A%2F%2Fraw.githubusercontent.com%2Ffadeevab%2Fmake.tmbundle%2F4f547858eac06033e562c1b08e3784466bd378fc%2FSyntaxes%2FMakefile.plist&grammar_text=&code_source=from-url&code_url=https%3A%2F%2Fraw.githubusercontent.com%2Fgithub%2Flinguist%2Fmaster%2Fsamples%2FMakefile%2Ffile-icons.make&code=). Still their tests don't contain the issues which are covered with my patch ($(info ...) at the beginning of the line and so on).

Commit:
Colorize when the expression stays at the beginning of line.

Commit:
Fix colorizing of built-in which stands after a common variable, for example, in the expression "var := $(var) $(info Hello!)" the "info" must be colorized. "variable" pattern is $$(....). Replace \G with (?<=\(), which means "get something right after the open brace which is found in the parent pattern". It makes the deal.

Commit:
Colorize (origin|flavor) at the same way as built-ins.

Commit:
Handle "\" at the end of prerequisites (the next line should be colorized as a continuation of previous). Previous pattern: search until TAB at the beginning of line. New pattern: search until the line without backslash "\" at the end. The "comment" pattern is fixed as well to continue colorizing until the end without a backslash. The patch is combined because comment could be placed in the recipe pattern with some circumstances.

fadeevab commented:
@infininight,
Finally, **I divided a pull request into 4 commits.** I hope it makes this pull request to be more clear. Thank you for review :+1:

fadeevab commented:
@infininight,
Thank you for review! :-)

fadeevab commented:
I believe one day this pull request will be submitted :)

@infininight, This is kindly reminder to review this pull request :+1: Thank you!

fadeevab commented:
Nobody loves makefile :(

fadeevab commented:
@sorbits, @alkemist, @bradchoate, @infininight, Hi, there! Any chances about this pull request?

Commit:
MIT license is added together with original TextMate license to support contribution to VS Code

Commit:
Allow spaces before ifeq, define, endif, endef, else.

Commit:
Handle braces properly inside the variable, e.g. $(shell ()()(()))

FadeevAlexandr committed:
Handle back-slash before end-of-line inside the $(variable).

FadeevAlexandr committed:
Handle "$$" and special variables.

Makefile use the following rule to parse variables:

 * A dollar sign followed by a character other than a dollar
   sign, open-parenthesis or open-brace treats that single
   character as the variable name.

$$, $@, $v, $*: highlight as a single variable.
$(): highlight as interpolation and parse content inside the braces.
$$(): "$$" is highlighted to show it's escaped, but parse content
      inside braces for convenient representation.

FadeevAlexandr committed:
Bugfix: "&lt" -> "&lt;"

infininight reviewed (changes requested):
In TextMate the `\G` assertion means the end of the parent match, in this case the end of the begin. So this commit does not actually change anything. The provided code sample works correctly in TextMate. I am unsure what level of support VSCode has for `\G` but I believe it if used in other grammars that VSCode uses?

infininight commented on Syntaxes/Makefile.plist:
```diff
@@ -64,7 +64,7 @@
 					</dict>
 					<key>end</key>
-					<string>\n</string>
+					<string>(?=[^\\])$</string>
```
This and the similar change below don't actually do anything. The look-ahead in ensuring there is not a `\` as the next character while then ensuring it is the end of the line. Both are always going to be true.

Also we do want to match the `\n` here to more proper scope the caret when it is at the end of the line since that is the default current scope (to the right of the caret).

fadeevab replied:
You are right. I had a trouble to continue comment which was started inside prerequisites so I tried to continue until I find "[^\\]\n". But here is a mistake definitely (it does nothing as you said). I should fix and re-test.

fadeevab replied:
About current status. Unfortunately, when I return \n instead of (?=\\)$ different bad things happen. I try to find why, but unfortunately, I struggled a lot previously, and now I'm not sure now I'll find a solution instead quickly. For example:

.PHONY: all
all: prereq # <------- "all:" looses the highlighting, or prereqs loose colorizing, etc.

**Test commit**: https://github.com/fadeevab/make.tmbundle/commit/db317ee2873264b96ab6f234683bee3b6324a0f0#diff-a941b8105726ed2f6d1ddbb7e4c9e996
**Result:** https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_url=https%3A%2F%2Fraw.githubusercontent.com%2Ffadeevab%2Fmake.tmbundle%2Fdb317ee2873264b96ab6f234683bee3b6324a0f0%2FSyntaxes%2FMakefile.plist&grammar_text=&code_source=from-url&code_url=https%3A%2F%2Fraw.githubusercontent.com%2Ffadeevab%2Fvscode%2Fmaster%2Fextensions%2Fmake%2Ftest%2Fcolorize-fixtures%2Fmakefile&code=

infininight commented on Syntaxes/Makefile.plist:
```diff
@@ -489,7 +489,7 @@
 			</dict>
 			<key>end</key>
-			<string>^(?!\t)</string>
+			<string>[^\\]$</string>
```
With this change added in I get the `meta.scope.target` ending before the `meta.scope.recipe` begins. The logic for extending the pre-requisite section should be contained in `meta.scope.prerequisites` rather than it's parent rule as that is what is really the cause here.

fadeevab replied:
I'll check it.

fadeevab replied:
Really, with my change target ending before the recipe begins. But in previous version the worse thing was happen: target ended with the first TAB inside the recipe.

Actually, I didn't change a lot previous logic, but I handled the following issue: TAB INSIDE MULTILINE PREREQUISITES also meant the end of prerequisites and begin of recipe: it's wrong in a case of multiline preresuisites ("\\"+"\n"): in a case of multiline prerequisites TAB isn't treat as a start of recipe.

infininight replied:
I have rewritten how prerequisites are matched entirely, should be corrected now.

infininight reviewed (commented):
OK, I have a commit to replace this which corrects the issue. Will skip this commit.

infininight reviewed (changes requested):
We are not interested in changing the license on one of our bundles, we would rather keep them uniform except for where we weren't the bundles originator.

This should not be required for VSCode as they use other bundles from TextMate with the same license.

Do you give permission for your changes to remain under the original license?

fadeevab commented:
You can skip the license change, because it was a request to me from VSCode team to make my repo to be upstream temporarily before you get to be active again and accept my pull request.

fadeevab commented:
@infininight , About \\G: I understand it should work like you said, but the next case didn't worked:

$(info ...) $(var) $(info ...)

The tailing $(info ...) isn't coloured. Changing \\G to the (?<=\\() makes the deal. As I remember it didn't work in TextMate as well as in VSCode

infininight reviewed (commented):
Can you describe this change? I get matching the escaped line but don't see anything in the docs about it ending the variable/function call.

fadeevab commented:
@infininight , Are you asking about "Handle back-slash before end-of-line inside the $(variable)."? Only  refefernse I can find: https://www.gnu.org/software/make/manual/html_node/Splitting-Lines.html

Makefile mechanically treats "\\+\n" as a space and concatenates the lines. I tested: no matter where is the back-slash is specified - inside the $(multiline-variable-with-backslashs ... ) or "var:=<mulitline value with backslashes>" or "goal: multiline-prerequisites" or anything else.

infininight commented:
OK, I think I've pulled in locally the changes for here that make sense. But a couple when devising some examples revealed deeper issues. Going to work on some general rewrites over the next few days and will put something up for you to test.

fadeevab commented:
Up to you, but I strictly recommend you to use my test above and compare with result of my master branch: even if my XML is not perfect from your point of view, it colorizes the test cases as I expect. Thank you for your review!

Commit:
Support of ${} variables.

I've had to add 2 duplicates of variables definition to properly catch $() and ${}.

The issue is to be in a catched context: $() or ${}. Common regex aka [)}] doesn't work: it incorrectly catches the cases $(}, or ${), or $(var}xxx), and other complex examples.
So, code duplication is the worst thing ever, but seems like there are no better ways currently in TextMate + makefile situation.

Commit:
Remove colorization after $$ at all.

Commit:
Recognize '!=' as a variable-assignment operator (issue #2)

Commit:
Fix of the variable assignment in a case of no spaces between a variable and value ("var:=value").

Problem: In the sentense kind of "variable:=value" the symbol ':' is highlited as a part of "variable:".
Cause: [^\s]+ is catched down to '=', dropping the ':=', '?=' and other cases.
Measure: restrict '=' to be a finisher of [^\s]+ in a case of [?:+!] behind one, allowing to select other assign operators.

Commit:
The pattern "recipe" is renamed to "target".

Cause: Preparing to add the real "recipe" pattern which starts from the TAB symbol.
Issue: #1.

Commit:
Correct syntax handling of a hashtag (#) symbols in the recipe.

Problem: In the recipe "<tab>echo '#'message" hastag and message behind is highlighted as a start of a comment.
Cause: 'recipe' patterns are not enabled.
Measure: Create a 'recipe' pattern starting with a [\t] symbol. Remove inactive matching.
Issue: #1.

Commit:
Removed shell interpolation inside backquotes ``.

Makefile syntax doesn't imply shell inside backquotes ``.

Commit:
Remove shell interpolation (coloring) in the recipe.

Problem:
  1. 'source.shell' hides coloring of Makefile variables, which must have a higher priority;
  2. incorrect coloring when there is no space before a shell: (a) <tab>@echo Test, (b) var!=echo 134
Cause: source.shell catches the context, it doesn't know about Makefile syntax.
Measure: Remove source.shell.

Commit:
@, - and + handling as prefixes to recipe lines.

@ suppresses the normal 'echo' of the command that is executed.
- ignores the exit status of the command that is execured.
+ means to execute command always, even under `make -n`.

Commit:
Handling a variable expansion in a variable name of assignment: var-$(nested-var) := value

Commit:
Bugfix of " @, - and + handling".

Problem: In a recipe "@./some-program" the "." (point) and "/" (slash) are colorized.
Cause: regex [@-+] means to colorize "from @ to +".
Measure: regex is fixed to [@\-+] to colorize only @,- and +.

Test case would be added to Microsoft/vscode.

tendrils committed:
Fix syntax highlighting for nested directives inside define blocks

fadeevab committed:
Merge pull request #3 from eqs-nz/master

Fix syntax highlighting for nested directives inside define blocks.

While it's not a widely used feature of Make, it is valid to nest define/endef blocks within one another arbitrarily in a makefile (at least in the case of GNU Make). This syntax change fixes highlighting of such nested blocks.

tbhartman committed:
Fix syntax highlighting of comments with prefixes

When a recipe contains a line with only a comment, and a prefix
precedes the comment, the highlighting is invalid (the comment is not
highlighted as such).

This allows comments to appear in recipes.

Fixes #4

fadeevab committed:
Merge pull request #5 from tbhartman/master

Fix syntax highlighting of comments with prefixes

Commit:
Revert "Fix syntax highlighting of comments with prefixes"

This reverts commit 8b350e4522f97f5fe371dee38d9d9baba2d48e8f.

Commit:
':', '#', '=' and '\s' are excluded from variable definition according to GNU Make documentation.

It improves highlighting of statements with double '=' like the following:

var:=$(val:.c=.o)

"A variable name may be any sequence of characters not containing ‘:’, ‘#’, ‘=’, or whitespace."
See https://www.gnu.org/software/make/manual/html_node/Using-Variables.html.

krobelus committed:
Prioritize directives over targets

Below line used to be interpreted as Make target with the name "export a ?= b".
Treat it as directive since this is what Make does.

	export a ?= b:c

fadeevab committed:
Prioritize directives over targets: export a ?= b:c

Below line used to be interpreted as Make target with the name export a ?= b.
Treat it as directive since this is what Make does.

export a ?= b:c

Commit:
Fix definition of ifdef|ifndef removing a copy-paste.

It fixes "Wrong tokenization of 'else' statement for ifdef in Makefile #10".

fadeevab committed:
Remove highlighting of comments inside of `define` directive.

It fixes "Makefile syntax highlighting partially breaks when a # is in a string #9".

fadeevab committed:
Stop highlighting comments in the recipes.

It is becuase the comment in the recipe is a part of Shell syntax, not the part of Makefile syntax.

It fixes "Makefile: incorrect syntax handling of comments with prefix #4".

fadeevab committed:
Improve ifdef-else-endif highlighting for the `else` directive.

Now the following test case should work:

```make
ifdef $(var-in-ifdef) # Comment
  $(info Good, as expected) # Comment
else $(var-can-be-here) # "else" is going to be colozired
  $(error 'Else' is not expected here) # Comment
endif # Comment
```

fadeevab committed:
Add comma separator for function calls and conditions

fadeevab committed:
Escaping a comment with an odd count of backslashes (\#, \\\#)

Problem: "#" means a beginning of a comment in ANY place unless it's escaped

How it works:
var:=blah#actual_comment
var:=blah\#not_a_comment
var:=blah\\#actual_comment
var:=blah\\\#not_a_comment

asottile committed:
Match `debian/rules` files based on shebang (#13)

`debian/rules` files typically start with `#!/usr/bin/make -f` but have a nonstandard filename so they don't match `Makefile`, etc.

Co-authored-by: Alexander Fadeev <fadeevab.com@gmail.com>

fadeevab committed:
A little remark

fadeevab committed:
Add support of `else ifeq|ifneq|...` conditions
